File: Java/word-break.java
//  Word Break

// Time O(N^2), Space O(N)
class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        int size = s.length();
        boolean[] words = new boolean[size+1];
        words[0] = true;
        for (int i=1; i<=size; i++) {
            for (int j=0; j<i; j++) {
                if (words[j] && wordDict.contains(s.substring(j, i))) {
                    words[i] = true;
                    break;
                }
            }
        }
        return words[size];
    }
}

File: leetcode/word-break.cc
//  Word Break

class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        vector<bool> dp(s.length()+1, false);
        dp[0] = true;
        for (int i = 0; i < s.length(); i ++) {
            for (int j = 0; j <= i; j ++) {
                string temp = s.substr(j, i-j+1);
                if (dp[j] && find(wordDict.begin(), wordDict.end(), temp)!=wordDict.end()) {
                    dp[i+1] = true;
                    break;
                }
            }
        }
        return dp[s.length()];
    }
};
